File: go/2022/day11/day11.go
package main

import (
	"fmt"
	"sort"
)

func main() {

	fmt.Printf("AoC 2022, Day9, Part1 solution is: %d \n", part1(true))
	fmt.Printf("AoC 2022, Day9, Part2 solution is: %d \n", part2(true))
}

func part1(useRealInput bool) int {
	monkeys := initInput()
	if !useRealInput {
		monkeys = intExample()
	}

	inspectedCounts := make([]int, len(monkeys))
	for round := 0; round < 20; round++ {
		for i, monkey := range monkeys {
			for _, item := range monkey.items {
				newItemValue := monkey.operation(item) / 3
				if newItemValue%monkey.testDivisibleBy == 0 {
					monkeys[monkey.trueMonkey].items = append(monkeys[monkey.trueMonkey].items, newItemValue)
				} else {
					monkeys[monkey.falseMonkey].items = append(monkeys[monkey.falseMonkey].items, newItemValue)
				}
			}
			inspectedCounts[i] += len(monkey.items)
			monkeys[i].items = []int{}
		}
	}

	sort.Ints(inspectedCounts)

	return inspectedCounts[len(inspectedCounts)-1] * inspectedCounts[len(inspectedCounts)-2]
}

func part2(useRealInput bool) int {
	monkeys := initInput()
	if !useRealInput {
		monkeys = intExample()
	}

	bigMod := 1
	for _, m := range monkeys {
		bigMod *= m.testDivisibleBy
	}

	inspectedCounts := make([]int, len(monkeys))
	for round := 0; round < 10000; round++ {
		for i, monkey := range monkeys {
			for _, item := range monkey.items {
				newItemVal := monkey.operation(item)
				newItemVal %= bigMod

				if newItemVal%monkey.testDivisibleBy == 0 {
					monkeys[monkey.trueMonkey].items = append(
						monkeys[monkey.trueMonkey].items, newItemVal)
				} else {
					monkeys[monkey.falseMonkey].items = append(
						monkeys[monkey.falseMonkey].items, newItemVal)
				}

			}
			inspectedCounts[i] += len(monkey.items)
			monkeys[i].items = []int{}
		}
	}

	sort.Ints(inspectedCounts)
	return inspectedCounts[len(inspectedCounts)-1] * inspectedCounts[len(inspectedCounts)-2]
}

type monkey struct {
	items                   []int
	operation               func(int) int
	testDivisibleBy         int
	trueMonkey, falseMonkey int
}

func intExample() []monkey {
	return []monkey{
		{
			items: []int{79, 98},
			operation: func(run int) int {
				return run * 19
			},
			testDivisibleBy: 23,
			trueMonkey:      2,
			falseMonkey:     3,
		},
		{
			items: []int{54, 65, 75, 74},
			operation: func(num int) int {
				return num + 6
			},
			testDivisibleBy: 19,
			trueMonkey:      2,
			falseMonkey:     0,
		},
		{
			items: []int{79, 60, 97},
			operation: func(num int) int {
				return num * num
			},
			testDivisibleBy: 13,
			trueMonkey:      1,
			falseMonkey:     3,
		},
		{
			items: []int{74},
			operation: func(num int) int {
				return num + 3
			},
			testDivisibleBy: 17,
			trueMonkey:      0,
			falseMonkey:     1,
		},
	}
}

func initInput() []monkey {
	return []monkey{
		{
			items: []int{77, 69, 76, 77, 50, 58},
			operation: func(run int) int {
				return run * 11
			},
			testDivisibleBy: 5,
			trueMonkey:      1,
			falseMonkey:     5,
		},
		{
			items: []int{75, 70, 82, 83, 96, 64, 62},
			operation: func(run int) int {
				return run + 8
			},
			testDivisibleBy: 17,
			trueMonkey:      5,
			falseMonkey:     6,
		},
		{
			items: []int{53},
			operation: func(run int) int {
				return run * 3
			},
			testDivisibleBy: 2,
			trueMonkey:      0,
			falseMonkey:     7,
		},
		{
			items: []int{85, 64, 93, 64, 99},
			operation: func(run int) int {
				return run + 4
			},
			testDivisibleBy: 7,
			trueMonkey:      7,
			falseMonkey:     2,
		},
		{
			items: []int{61, 92, 71},
			operation: func(run int) int {
				return run * run
			},
			testDivisibleBy: 3,
			trueMonkey:      2,
			falseMonkey:     3,
		},
		{
			items: []int{79, 73, 50, 90},
			operation: func(run int) int {
				return run + 2
			},
			testDivisibleBy: 11,
			trueMonkey:      4,
			falseMonkey:     6,
		},
		{
			items: []int{50, 89},
			operation: func(run int) int {
				return run + 3
			},
			testDivisibleBy: 13,
			trueMonkey:      4,
			falseMonkey:     3,
		},
		{
			items: []int{83, 56, 64, 58, 93, 91, 56, 65},
			operation: func(run int) int {
				return run + 5
			},
			testDivisibleBy: 19,
			trueMonkey:      1,
			falseMonkey:     0,
		},
	}
}

File: python/2022/day11/day11.py
from dataclasses import dataclass


@dataclass
class Monkey:
    items: list[int]
    operation: str
    test_factor: int
    true_monkey_target: int
    false_monkey_target: int
    items_check_count: int = 0

    def update_item_check_counter(self):
        self.items_check_count += len(self.items)


def part1(input: str) -> int:
    monkeys = parse_input(input)
    return run(monkeys=monkeys, iterations=20, fn=lambda x: x // 3)


def part2(input: str) -> int:
    monkeys = parse_input(input)

    mod_value = 1
    for monkey in monkeys:
        mod_value *= monkey.test_factor

    return run(monkeys=monkeys, iterations=10000, fn=lambda x: x % mod_value)


def run(monkeys: list[Monkey], iterations: int, fn) -> int:
    for _ in range(iterations):
        for monkey in monkeys:
            for item in monkey.items:
                item = eval(monkey.operation)(item)
                item = fn(item)
                if item % monkey.test_factor == 0:
                    monkeys[monkey.true_monkey_target].items.append(item)
                else:
                    monkeys[monkey.false_monkey_target].items.append(item)
            monkey.update_item_check_counter()
            monkey.items = []

        # print_stats(monkeys)

    items_count = [monkey.items_check_count for monkey in monkeys]
    items_count.sort()
    ans = items_count[-1] * items_count[-2]
    return ans


def parse_input(input: str) -> list[Monkey]:

    monkeys = []

    for block in read_input(input):
        lines = block.split("\n")
        monkey_items = list(map(int, lines[1].split(": ")[1].split(", ")))
        monkey_operation = "lambda old: " + lines[2].split(" = ")[1]
        monkey_divisible_value = int(lines[3].split(" ")[-1])
        monkey_true_id = int(lines[4].split(" ")[-1])
        monkey_false_id = int(lines[5].split(" ")[-1])

        monkey = Monkey(
            items=monkey_items,
            operation=monkey_operation,
            test_factor=monkey_divisible_value,
            true_monkey_target=monkey_true_id,
            false_monkey_target=monkey_false_id,
        )

        monkeys.append(monkey)

    return monkeys


def read_input(input: str) -> list[str]:
    with open(input) as fn:
        return fn.read().strip().split("\n\n")


def print_stats(monkeys: list[Monkey]):
    for i, monkey in enumerate(monkeys):
        print(f"Monkey-{i}: {monkey.items}, items-count: {monkey.items_check_count}")
    print("======================")


def main():
    puzzle_input_path = "2022/day11/input.txt"
    print(f"AoC2022 Day10, Part1 solution is: {part1(input=puzzle_input_path)}")
    print(f"AoC2022 Day10, Part2 solution is: {part2(input=puzzle_input_path)}")


if __name__ == "__main__":
    main()
